### PR TITLE
added category model

### PR DESCRIPTION
**Pull request by Oleksandr-Tymoshenko**

### CODE REVIEW

File: src/main/java/book/store/onlinebookstore/service/impl/BookServiceImpl.java
Comment (mmosii) [inline src/main/java/book/store/onlinebookstore/service/impl/BookServiceImpl.java:1]: I as understand requirements, you have to refactor your BookService so he can save or update book with categoryIds, when you pass them with createBookRequestDto.  I don't see any ways to set categories to your books now.  Maybe i am wrong, not an expert)

**Review by mmosii — CHANGES_REQUESTED** [review]
Nice job! have one suggestion

File: src/main/java/book/store/onlinebookstore/service/impl/BookServiceImpl.java
Comment (Oleksandr-Tymoshenko) [reply src/main/java/book/store/onlinebookstore/service/impl/BookServiceImpl.java:1]: As i undersand, categories are setted by updating or creating book. I added field categories to CreateBookRequestDto, so admin can set category if he wants to

File: src/main/java/book/store/onlinebookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -36,7 +36,10 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
                 .csrf(AbstractHttpConfigurer::disable)
                 .authorizeHttpRequests(
                         auth -> auth
-                                .requestMatchers("/api/auth/**")
+                                .requestMatchers("/api/auth/**",
+                                        "/error",
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/config/SecurityConfig.java:40]: what is this path for?

File: src/main/java/book/store/onlinebookstore/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @Operation(summary = "Update category by id",
+            description = "Update the category by its id number")
+    public CategoryDto updateCategory(@PathVariable @Positive Long id,
+                                      @RequestBody @Valid CreateCategoryRequestDto requestDto) {
+        return categoryService.update(id, requestDto);
+    }
+
+    @DeleteMapping("/{id}")
+    @ResponseStatus(HttpStatus.NOT_FOUND)
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/controller/CategoryController.java]: NOT_FOUND ? O_o

File: src/main/java/book/store/onlinebookstore/dto/book/BookDto.java [hunk 3]
```diff
@@ -1,13 +1,17 @@
...
-public record BookDto(Long id,
-                      String title,
-                      String author,
-                      String isbn,
-                      BigDecimal price,
-                      String description,
-                      String coverImage) {
-
+@Data
+public class BookDto {
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/dto/book/BookDto.java:8]: why have you changed it to class?

File: src/main/java/book/store/onlinebookstore/dto/book/CreateBookRequestDto.java [hunk 4]
```diff
@@ -5,12 +5,14 @@
...
 import jakarta.validation.constraints.Positive;
 import java.math.BigDecimal;
+import java.util.Set;
 
 public record CreateBookRequestDto(
         @NotBlank String title,
         @NotBlank String author,
         @NotBlank @Pattern(regexp = "^(?=(?:\\D*\\d){10}(?:(?:\\D*\\d){3})?$)[\\d-]+$") String isbn,
         @NotNull @Positive BigDecimal price,
+        Set<Long> categoriesIds,
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/dto/book/CreateBookRequestDto.java]: you can also validate categories:
```
@NotEmpty Set<@Positive Long> categoriesIds
```

File: src/main/java/book/store/onlinebookstore/exception/GlobalExceptionHandler.java [hunk 5]
```diff
@@ -48,6 +49,11 @@ public ResponseEntity<String> handleAccessDenied(AccessDeniedException ex) {
         return new ResponseEntity<>(ex.getMessage(), HttpStatus.FORBIDDEN);
     }
 
+    @ExceptionHandler(JwtException.class)
+    public ResponseEntity<String> handleJwtException(JwtException ex) {
+        return new ResponseEntity<>(ex.getMessage(), HttpStatus.FORBIDDEN);
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/exception/GlobalExceptionHandler.java]: well it depends on what stays behind `JwtException`

if you failed to parse and validate JWT - then HTTP 401 (UNAUTHORIZED) should be returned

if parsing and validation succeeded, but JWT is missing some grants - then HTTP 403 (FORBIDDEN) should be returned

File: src/main/java/book/store/onlinebookstore/repository/book/BookRepository.java [hunk 6]
```diff
@@ -1,8 +1,20 @@
...
 import book.store.onlinebookstore.model.Book;
+import java.util.List;
+import java.util.Optional;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.EntityGraph;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/repository/book/BookRepository.java:14]: `@Repository` is missing

File: src/main/java/book/store/onlinebookstore/repository/book/BookRepository.java [hunk 7]
```diff
@@ -1,8 +1,20 @@
...
+    @EntityGraph(attributePaths = "categories")
+    Optional<Book> findById(Long id);
+
+    @EntityGraph(attributePaths = "categories")
+    Page<Book> findAll(Pageable pageable);
+
+    List<Book> findAllByCategoriesId(Long id, Pageable pageable);
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/repository/book/BookRepository.java:21]: will this work at all? have you tested it?

File: src/main/java/book/store/onlinebookstore/repository/category/CategoryRepository.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package book.store.onlinebookstore.repository.category;
+
+import book.store.onlinebookstore.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/repository/category/CategoryRepository.java:8]: `@Repository` is missing

File: src/main/java/book/store/onlinebookstore/repository/category/CategoryRepository.java [hunk 9]
```diff
@@ -0,0 +1,8 @@
+package book.store.onlinebookstore.repository.category;
+
+import book.store.onlinebookstore.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/repository/category/CategoryRepository.java]: redundant blank line

File: src/main/java/book/store/onlinebookstore/service/impl/BookServiceImpl.java [hunk 10]
```diff
@@ -33,6 +34,12 @@ public BookDto findById(Long id) {
         return bookMapper.toDto(book);
     }
 
+    @Override
+    public List<BookDtoWithoutCategoryIds> findBooksByCategoryId(Long id, Pageable pageable) {
+        return bookRepository.findAllByCategoriesId(id, pageable).stream()
+                .map(bookMapper::toDtoWithoutCategories).toList();
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/service/impl/BookServiceImpl.java]:
```suggestion
        return bookRepository.findAllByCategoriesId(id, pageable).stream()
                .map(bookMapper::toDtoWithoutCategories)
                .toList();

```

File: src/main/java/book/store/onlinebookstore/service/impl/CategoryServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class CategoryServiceImpl implements CategoryService {
+    private final CategoryRepository categoryRepository;
+    private final CategoryMapper categoryMapper;
+
+    @Override
+    public List<CategoryDto> findAll(Pageable pageable) {
+        return categoryRepository.findAll().stream().map(categoryMapper::toDto).toList();
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/service/impl/CategoryServiceImpl.java]:
```suggestion
        return categoryRepository.findAll().stream()
                      .map(categoryMapper::toDto)
                      .toList();

```

File: src/main/java/book/store/onlinebookstore/service/impl/CategoryServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    public CategoryDto getById(Long id) {
+        return categoryMapper.toDto(categoryRepository
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/service/impl/CategoryServiceImpl.java]: let's make it more readable:
```
return repo.findById(id)
              .map(mapper::toDto)
              .orElseThrow(...);
```

File: src/main/java/book/store/onlinebookstore/service/impl/CategoryServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+                .findById(id)
+                .orElseThrow(() -> new EntityNotFoundException("Category was not fount by id "
+                        + id)));
+    }
+
+    @Override
+    public CategoryDto save(CreateCategoryRequestDto categoryRequestDto) {
+        return categoryMapper.toDto(categoryRepository
+                .save(categoryMapper.toCategory(categoryRequestDto)));
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/service/impl/CategoryServiceImpl.java]: this is an example, when additional variable would add up to readability:
```suggestion
        Category category = categoryMapper.toCategory(categoryRequestDto);
        return categoryMapper.toDto(categoryRepository.save(category));

```

File: src/main/java/book/store/onlinebookstore/service/impl/CategoryServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    public CategoryDto update(Long id, CreateCategoryRequestDto categoryRequestDto) {
+        Category category = categoryMapper.toCategory(categoryRequestDto);
+        category.setId(id);
+        return categoryMapper.toDto(categoryRepository.save(category));
```
Comment (NazarSmith228) [inline src/main/java/book/store/onlinebookstore/service/impl/CategoryServiceImpl.java]: the same mistake as in book update some time ago

you should load the category by id, then update it (via Mapstruct), and then save

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/book/store/onlinebookstore/config/SecurityConfig.java (hunk 1 above)
Comment (Oleksandr-Tymoshenko) [reply src/main/java/book/store/onlinebookstore/config/SecurityConfig.java:40]: without it server responses to a USER_ROLE who wants to create/update/delete with 401 status, with it with 403

File: src/main/java/book/store/onlinebookstore/repository/book/BookRepository.java (hunk 7 above)
Comment (Oleksandr-Tymoshenko) [reply src/main/java/book/store/onlinebookstore/repository/book/BookRepository.java:21]: yes

File: src/main/java/book/store/onlinebookstore/dto/book/UpdateBookRequestDto.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+
+public record UpdateBookRequestDto(
+        String title,
+        String author,
+        @Pattern(regexp = "^(?=(?:\\D*\\d){10}(?:(?:\\D*\\d){3})?$)[\\d-]+$") String isbn,
+        @Positive BigDecimal price,
+        Set<Long> categoriesIds,
+        String description, String coverImage) {
+
+}
```
Comment (liliia-ponomarenko) [inline src/main/java/book/store/onlinebookstore/dto/book/UpdateBookRequestDto.java]:
```suggestion
        String description, 
        String coverImage) {
}
```

**Review by liliia-ponomarenko — APPROVED** [review]
Great! 😊

**Pull request merged**